### PR TITLE
Requirements: Add installation methods

### PR DESCRIPTION
**Pull request by Adrianzatreanu**

### CODE REVIEW

File: coalib/bears/requirements/PipRequirement.py [hunk 1]
```diff
@@ -24,3 +25,34 @@ def __init__(self, package, version=""):
...
+    def is_installed(self):
+        """
+        Checks if the dependency is installed.
+
+        >>> pr = PipRequirement('pip')
+        >>> pr.is_installed()
+        True
+
+        >>> pr = PipRequirement('some_package')
+        >>> pr.is_installed()
```
Comment (sils1297) [inline coalib/bears/requirements/PipRequirement.py]: shorten those to `PipRequirement('...').is_installed()`

File: coalib/bears/requirements/PipRequirement.py [hunk 2]
```diff
@@ -24,3 +25,34 @@ def __init__(self, package, version=""):
...
+        False
+
+        :param return: True if dependency is installed, false otherwise.
+        """
+        if shutil.which(self.package):
```
Comment (sils1297) [inline coalib/bears/requirements/PipRequirement.py]: nah that doesn't work for python, not all packages ship with a binary! E.g. coala-bears won't have a `coala-bears` binary registered and thus shutil.which can't find it

**Comment by sils1297** [comment]
unack https://github.com/coala-analyzer/coala/pull/2393/commits/def51ddbbc5eb3ff18d8518d49acb5f4cb8aec2a

File: coalib/bears/requirements/NpmRequirement.py [hunk 3]
```diff
@@ -24,3 +25,34 @@ def __init__(self, package, version=""):
         :param version: A version string. Leave empty to specify latest version.
         """
         PackageRequirement.__init__(self, 'npm', package, version)
+
+    def install_command(self):
+        """
+        Creates the installation command for the instance of the class.
+
+        >>> pr = NpmRequirement('alex', '2')
+        >>> pr.install_command()
```
Comment (sils1297) [inline coalib/bears/requirements/NpmRequirement.py]: shorten

File: coalib/bears/requirements/PipRequirement.py (hunk 2 above)
Comment (Adrianzatreanu) [inline coalib/bears/requirements/PipRequirement.py]: alright, for python ill go with ``pip show`` then

File: coalib/bears/requirements/NpmRequirement.py [hunk 4]
```diff
@@ -24,3 +25,34 @@ def __init__(self, package, version=""):
...
+        >>> pr.is_installed()
+        True
+
+        >>> pr = NpmRequirement('some_package')
+        >>> pr.is_installed()
+        False
+
+        :param return: True if dependency is installed, false otherwise.
+        """
+        if shutil.which(self.package):
```
Comment (sils1297) [inline coalib/bears/requirements/NpmRequirement.py]: same here, rather ask npm for it

Comment (Adrianzatreanu) [inline coalib/bears/requirements/NpmRequirement.py]: i dont know how.. i havent found a way to ask npm to show me if its installed..ideas?

**Comment by Adrianzatreanu** [comment]
CC @justuswilhelm as this PR should interest you a little :D

File: coalib/bears/requirements/PipRequirement.py [hunk 5]
```diff
@@ -24,3 +25,29 @@ def __init__(self, package, version=""):
...
+        >>> PipRequirement('pip').is_installed()
+        True
+
+        >>> PipRequirement('some_package').is_installed()
+        False
+
+        :param return: True if dependency is installed, false otherwise.
+        """
+        return not subprocess.call(['pip', 'show', self.package],
+                                   stdout=subprocess.DEVNULL)
```
Comment (sils1297) [inline coalib/bears/requirements/PipRequirement.py]: what about stderr? Maybe we should put `call_without_output = functools.partial(subprocess.call, stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL` in `Shell.py`?

File: tests/bears/requirements/RequirementsTest.py [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+    def setUp(self):
+        self.GemRequirement = GemRequirement('ruby')
+        self.GemRequirementBad = PipRequirement('some_bad_package')
+
+    unittest.skipIf(gem_check(), "Ruby is not installed.")
+
+    def test_InstalledRequirement(self):
+        self.assertEqual(self.GemRequirement.is_installed(), True)
+
+    unittest.skipIf(gem_check(), "Ruby is not installed.")
```
Comment (justuswilhelm) [inline tests/bears/requirements/RequirementsTest.py]: Accidental non-decorator

File: coalib/bears/requirements/PipRequirement.py (hunk 5 above)
Comment (Adrianzatreanu) [inline coalib/bears/requirements/PipRequirement.py]: Why do we care about stderr? its not shown on the screen. I've placed it on the Npm one, as there it is.

File: tests/bears/requirements/RequirementsTest.py (hunk 6 above)
Comment (Adrianzatreanu) [inline tests/bears/requirements/RequirementsTest.py]: ah correct!

**Comment by Adrianzatreanu** [comment]
im awaiting a full review and then I'll make the changes. thanks, guys :) you the best!

File: coalib/bears/requirements/PipRequirement.py (hunk 5 above)
Comment (sils1297) [inline coalib/bears/requirements/PipRequirement.py]: who knows, let's do the functools partial thing

**Comment by sils1297** [comment]
unack https://github.com/coala-analyzer/coala/pull/2393/commits/0a9129d41227c00fa2a091c9105f64f98ec04b2f and the others for I want the functools partial thing to have it clean in the first place.

**Comment by Adrianzatreanu** [comment]
Updated PR with @sils1297 mighty callwithoutoutput and @justuswilhelm notified missing decorators!

**Comment by Adrianzatreanu** [comment]
looking for help, not sure why the tests fail on CircleCI, the tests work fine locally @justuswilhelm @sils1297 @Makman2

**Comment by justuswilhelm** [comment]
Seems there is an issue with skipping this test
```
    @unittest.skipIf(shutil.which('gem') is None, "Ruby is not installed.")
    def test_InstalledRequirement(self):
>       self.assertEqual(self.GemRequirement.is_installed(), True)
E       AssertionError: False != True
```
Maybe shutil.which is not giving you the right result on this machine? When in doubt use `/usr/bin/type`

**Comment by Adrianzatreanu** [comment]
but ``type`` wouldnt work on windows, would it? @justuswilhelm

File: coalib/bears/requirements/GemRequirement.py [hunk 7]
```diff
@@ -29,3 +30,28 @@ def __init__(self, package, version="", require=""):
...
+                                                          self.require)
+        return "gem install {}".format(self.package)
+
+    def is_installed(self):
+        """
+        Checks if the dependency is installed.
+
+        :param return: True if dependency is installed, false otherwise.
+        """
+        return not call_without_output(['gem', 'list', '-i', self.package])
```
Comment (Makman2) [inline coalib/bears/requirements/GemRequirement.py]: use a tuple :+1:

File: tests/bears/requirements/RequirementsTest.py [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+    @unittest.skipIf(shutil.which('npm') is None, "Npm is not installed.")
+    def test_NotInstalledRequirement(self):
+        self.assertEqual(self.NpmRequirementBad.is_installed(), False)
+
+
+class GemRequirementTestCase(unittest.TestCase):
+
+    def setUp(self):
+        self.GemRequirement = GemRequirement('ruby')
+        self.GemRequirementBad = PipRequirement('some_bad_package')
```
Comment (Makman2) [inline tests/bears/requirements/RequirementsTest.py]: not `GemRequirement`?

File: tests/bears/requirements/RequirementsTest.py [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+
+    @unittest.skipIf(shutil.which('gem') is None, "Ruby is not installed.")
+    def test_InstalledRequirement(self):
+        self.assertEqual(self.GemRequirement.is_installed(), True)
+
+    @unittest.skipIf(shutil.which('gem') is None, "Ruby is not installed.")
+    def test_NotInstalledRequirement(self):
+        self.assertEqual(self.GemRequirementBad.is_installed(), False)
```
Comment (Makman2) [inline tests/bears/requirements/RequirementsTest.py]: could you please move each test class to the according commit which introduces the specific install methods? if we are going to revert stuff (just hypothetically) to the commits before this one, we have a coverage drop :+1:

Comment (Adrianzatreanu) [inline tests/bears/requirements/RequirementsTest.py]: alright

**Comment by Adrianzatreanu** [comment]
moved to https://github.com/coala-analyzer/coala/pull/2397 due to git corruptions

**Pull request closed without merging**